s3fs/core.py

```diff
@@ -888,6 +889,62 @@ async def _get_file(self, rpath, lpath, version_id=None):
         finally:
             body.close()
 
+    async def _simple_info(self, path):
+        bucket, key, _ = self.split_path(path)
+        prefix = key.strip("/")
+
+        out = await self._call_s3(
+            "list_objects_v2",
+            self.kwargs,
+            Bucket=bucket,
+            Prefix=prefix,
+            Delimiter="/",
+            MaxKeys=1,
+            **self.req_kw,
+        )
+        [file] = out.get("Contents", [None])
+        if file and file["Key"] == prefix:
+            return {
+                "ETag": file["ETag"],
+                "Key": "/".join([bucket, key]),
+                "LastModified": file["LastModified"],
+                "Size": file["Size"],
+                "size": file["Size"],
+                "name": "/".join([bucket, key]),
+                "type": "file",
+                "StorageClass": "STANDARD",
+                "VersionId": file.get("VersionId"),
+            }
+        elif not (file or out.get("KeyCount", 0) or out.get("CommonPrefixes", [])):
+            raise FileNotFoundError(path)
```

Comment:
Aren't there `else` cases here:
- we got a file by a different name, so the original file was not found
- we got prefix(es), so the path is a directory (does MaxKeys restrict the number of prefixes? Why use delimiter in the call at all?)
- we got a path which is the same except for an extra "/"

What happens if we fail all the ifs, and return `None`?

Reply:
> What happens if we fail all the ifs, and return None?

The `else: ` is the condition where the prefix doesn't exist as is, and there are other prefixes. For example for a directory like this;
```
very/
   a12/
      another
   a1234
```

`very/a123` will be handled in the first condition, `very/bsomething` will be handled in the second condition since no such prefix exists either as a file or a directory and `very/a1` and `very/a2` will fallback to the directory search which would return true for `very/a2` and raise file not found error for `very/a`.

I couldn't find a test case which isn't handled by this logic

Reply:
No, else is a totally different path. For example there might other prefixes matched, so it might be a directory etc